### DISCUSSION TITLE
Starting `ragna` from `jhub-apps`

### DISCUSSION THREAD

**Discussion by aktech** [nebari-dev/jhub-apps#147] [Announcements]
This discussion is more of a documentation on how to use Ragna via `jhub-apps`:

The workflow for running Ragna on JHub Apps would be something like this:

- Create an app that will launch Ragna API, after creation you'll know the url of the deployed API app, this needs to be Public as the Ragna UI authenticates to the Ragna API in a headless way (which will trigger JupyterHub Authentication in non-public app before Ragna Authentication) and not on the browser
- Edit the `ragna.toml` to add `root_path` and update `url` (see sample `ragna.toml` below for example) and redeploy, so that API works in the deployed app (as the url for the going to be deployed app on jhub-apps is not known in advance)
- Create another app for Ragna UI app

Input for custom command field in jhub-apps for starting:

(You'd need to use the paths for your deployment/system)

## Ragna UI

```
ragna ui {--}config /Users/aktech/quansight/jhub-apps/ragna.toml {--}no-open-browser {--}no-start-api --destport=31477
```

## Ragna API

```
ragna api {--}config /Users/aktech/quansight/jhub-apps/ragna.toml --destport=31476
```

In the Environment variable field you'd need to pass in the 

```
# change the url here to match the url of deployed jhub-apps
{"BOKEH_ALLOW_WS_ORIGIN": "127.0.0.1:8000"}
```

## `ragna.toml`

Sample `ragna.toml` that worked locally:

```yml
# change to match your directory and same for all others below
local_root = "/Users/aktech/.cache/ragna"
authentication = "ragna.deploy.RagnaDemoAuthentication"
document = "ragna.core.LocalDocument"
source_storages = [
    "ragna.source_storages.RagnaDemoSourceStorage",
]
assistants = [
    "ragna.assistants.RagnaDemoAssistant",
]

[api]
hostname = "127.0.0.1"
port = "31476"

# change this url to match the url of the created ragna api app url
url = "http://127.0.0.1:8000/user/aktech/ragna-api-f8ed40f/"
origins = ["http://127.0.0.1:8000"]
database_url = "sqlite:////Users/aktech/.cache/ragna/ragna.db"
root_path = "/user/aktech/ragna-api-f8ed40f/"


[ui]
hostname = "127.0.0.1"
port = 31477
origins = [
    "http://127.0.0.1:31477",
]
```

**Comment by pmeier** [comment]
> * Edit the `ragna.toml` to add `root_path` and update `url` (see sample `ragna.toml` below for example) and redeploy, so that API works in the deployed app (as the url for the going to be deployed app on jhub-apps is not known in advance)

That could also be done by setting the `RAGNA_API_ROOT_PATH` and `RAGNA_API_URL` environment variables if that makes it any easier.

> In the Environment variable field you'd need to pass in the
> 
> ```
> # change the url here to match the url of deployed jhub-apps
> {"BOKEH_ALLOW_WS_ORIGIN": "127.0.0.1:8000"}
> ```

We should look into this. This is what the `ui.origins` configuration option is for. It should not need  to be required to set an environment variable for Bokeh.
